### PR TITLE
Issue #547: Added a Sliding Time Window implementation to measure met…

### PR DESCRIPTION
**Pull request by RobWin**
First draft of a Sliding Time Window metric store is pushed.

The implementation is backed by a sliding time window that aggregates only the calls made in the last N seconds.
The sliding time window is implemented with a circular array of N buckets.
If the time window size is 10 seconds, the circular array has always 10 buckets. Every bucket
stores the outcome of all calls which happen in a certain epoch second. The head bucket of the circular array stores the call outcomes of the current epoch second. The other buckets stored the call outcomes of the previous N-1 epoch seconds.

The sliding window does not store call outcomes (tuples) individually, but incrementally updates partial aggregations (bucket) and a total aggregation.
The total aggregation is updated incrementally when a new call outcome is recorded. When the oldest bucket is evicted, the partial aggregation of that bucket
is subtracted from the total aggregation and the bucket is reset. (Subtract-on-Evict)

The time to retrieve a Snapshot is constant 0(1), since the Snapshot is pre-aggregated and is independent of the time window size.
The space requirement (memory consumption) of this implementation should be nearly constant O(n), since the call outcomes (tuples) are not stored individually.
Only N partial aggregations and 1 total aggregation are created.

Feel free to review it ;)

This implementation would also allow us to track slow calls and adapt the CircuitBreaker so that it can transition to OPEN when the percentage of slow calls is above a certain threshold. It would solve #381

### CODE REVIEW

File: resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/Snapshot.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.core.metrics;
+
+import java.time.Duration;
+
```
Comment (Romeh) [inline resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/Snapshot.java]: Thinking loud here , Small question for discussion , can we maybe have Snapshot<T> and same for Metrics<T> and let the user define what type of APIs/counts he can use for metrics by call get T . (api call) or we should enforce what type of stats he can get and should follow that ?

File: resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/SlidingTimeWindowMetrics.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+ * The sliding window does not store call outcomes (tuples) individually, but incrementally updates partial aggregations (bucket) and a total aggregation.
+ * The total total aggregation is updated incrementally when a new call outcome is recorded. When the oldest bucket is evicted, the partial total aggregation of that bucket
+ * is subtracted from the total totalAggregation and the bucket is reset. (Subtract-on-Evict)
+ *
+ * The time to retrieve a Snapshot is constant 0(1), since the Snapshot is pre-aggregated and is independent of the time window size.
+ * The space requirement (memory consumption) of this implementation should be nearly constant O(n), since the call outcomes (tuples) are not stored individually.
+ * Only {@code N} partial aggregations and 1 total total aggregation are created.
+ */
+public class SlidingTimeWindowMetrics implements Metrics {
+
```
Comment (Romeh) [inline resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/SlidingTimeWindowMetrics.java]: I feel we should separate the sliding window logic and storage into separate class aside from the metrics , so anyone want to use outside the metrics Interface , he can do it , i am thinking about the adaptive bulkhead case here , what do u think ?

File: resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/SlidingWindowMetrics.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        latestMeasurement.reset();
+        return latestMeasurement;
+    }
+
+    /**
+     * Returns the head partial aggregation of the circular array.
+     *
+     * @return the head partial aggregation of the circular array
+     */
+    Measurement getLatestMeasurement(){
```
[the remark this answers is not in the document]
Comment (RobWin) [reply resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/SlidingWindowMetrics.java]: Fixed

File: resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/SnapshotImpl.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ *
+ */
+package io.github.resilience4j.core.metrics;
+
+import java.time.Duration;
+
+public class SnapshotImpl implements Snapshot {
+
+    private long totalDurationInMillis;
```
[the remark this answers is not in the document]
Comment (RobWin) [reply resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/SnapshotImpl.java]: Fixed

File: resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/SlidingWindowCircuitBreakerMetrics.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Records a call which was not permitted, because the CircuitBreaker state is OPEN.
+     */
+    void onCallNotPermitted() {
+        numberOfNotPermittedCalls.increment();
+    }
+
+    @Override
+    public float getFailureRate() {
```
[the remark this answers is not in the document]
Comment (RobWin) [reply resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/SlidingWindowCircuitBreakerMetrics.java]: This class is deleted. Ignore it please.

File: resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/Snapshot.java (hunk 1 above)
Comment (RobWin) [reply resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/Snapshot.java]: Can you give an example? I don't understand it

Comment (Romeh) [reply resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/Snapshot.java]: per discussion not applicable

**Comment by wl1244hotmai** [comment]
Another question is that, is there any way don't to record the rate of slow call? Sometimes we already had timeout settings in http connection, so we don't need it. 
After this change, I may need to configure the slow SLOW_CALL_DURATION_THRESHOLD to a big one to avoid trigger it. It there any way to disable it?

**Comment by RobWin** [comment]
Hi, could you please add your comments to the issue #547 instead of the the PR?

**Pull request merged**